pkg/sbom/nodelist.go

```diff
@@ -419,9 +418,100 @@ func (nl *NodeList) Equal(nl2 *NodeList) bool {
 	}
 
 	if !reflect.DeepEqual(nlNodes, nl2Nodes) {
-		logrus.Infof("NodeList 1: %+v\nNodeList 2: %+v", nlNodes, nl2Nodes)
+		logrus.Info("No: nodes")
 		return false
 	}
 
 	return true
 }
+
+// RelateNodeListAtID relates the top level nodes in nl2 to the node with ID
+// nodeID using a relationship of type edgeType. Returns an error if nodeID cannot
+// be found in the graph. This function assumes that nodes in nl and nl2 having
+// the same ID are equivalent and will be deduped.
+func (nl *NodeList) RelateNodeListAtID(nl2 *NodeList, nodeID string, edgeType Edge_Type) error {
+	// Check the node exists
+	nlIndex := nl.indexNodes()
+	nlEdges := nl.indexEdges()
+
+	if _, ok := nlIndex[nodeID]; !ok {
+		return fmt.Errorf("node with ID %s not found", nodeID)
+	}
+
+	// Check of we have edges matching
+	var edge *Edge
+	if _, ok := nlEdges[nodeID]; ok {
+		if _, ok2 := nlEdges[nodeID][edgeType]; ok2 {
+			edge = nlEdges[nodeID][edgeType][0]
+		}
+	}
+
+	if edge == nil {
+		edge = &Edge{
+			Type: edgeType,
+			From: nodeID,
+			To:   nl2.RootElements,
+		}
+		nl.Edges = append(nl.Edges, edge)
+	} else {
+		// Perhaps we should filter these
+		edge.To = append(edge.To, nl2.RootElements...)
+	}
+
+	for _, n := range nl2.Nodes {
+		if _, ok := nlIndex[n.Id]; ok {
+			continue
+		}
+		nl.AddNode(n)
+	}
+
+	return nil
+}
+
+// GetNodesByPurlType returns a nodelist containing all nodes that match
+// a purl (package url) type. An empty purlType returns a blank nodelist
+func (nl *NodeList) GetNodesByPurlType(purlType string) *NodeList {
+	ret := &NodeList{}
+	if nl == nil {
+		return ret
+	}
+
+	for _, n := range nl.Nodes {
+		// I think the SPDX libraries have a bug where an extra slash is added when parsing purls
```

Comment:
Should we put in a PR to the upstream SPDX libraries? Feel free to do that as part of this work.

Reply:
I think I will, I think it will take more time to find it than the actual fix :D